### DISCUSSION TITLE
Translations: Should technical terms be translated?

### DISCUSSION THREAD

**Discussion by deavid** [google/comprehensive-rust#1184] [General]
Just a bit of an open discussion to see what others think here.

I am noticing that by default we are attempting to translate almost everything into the target language. My experience is mainly with Spanish vs English so I'm not sure if my perspective applies to other languages too, but a friend also noticed the same on the German translation and we seem to share the same opinion.

In IT related stuff we tend to borrow words from English for technical stuff: borrowing, pattern matching, compile time, etc. These words while they were used because the original meaning kind of matched what we needed to convey, over time they start to convey specifically something different in the context. Even in English I bet it would sound weird if we start to use synonyms to replace these.

As a result, for Spanish speakers, we also borrow the same English words verbatim just for that special meaning. This is also extra useful for us as it points out very clearly that this is a technical specific term and not part of normal conversation. We tend to give it a special meaning. (For example, I thought time ago that "chat" was specific to text chat application, and I was puzzled when someone called me on the phone (voice) just after asking me to chat)

The usual way we deal with these English terms in Spanish is to have them in italics (or quotes if italics doesn't work) then add the meaning in Spanish in parenthesis afterwards.

Once the reader gets used to the new English term we just put them in italics and drop the Spanish counterpart.

On the other hand translating every term tends to feel... cringe? (in Spanish at least) But also very confusing: "Borrow checker" -> "Comprobador de préstamos". "Ownership" -> "Propiedad" / "Dueño". It does read as very unnatural and forced, but also the lack of the original English words threatens the ability to google it. See: https://www.google.com/search?client=firefox-b-e&q=rust+comprobador+de+pr%C3%A9stamos (5 results, some of the titles look like bad translations)

It is also true that in Spain, the RAE (Spanish Language Academy) has a push for "translating everything", although most people dislike this approach in practice. But they also recommend the approach I outlined above for foreign words.

I'm not sure how other languages deal with this. What does everyone think? Is this something we would like to have a common way to deal with across languages, or should it be per-language? What is your opinion?

**Comment by domenukk** [comment]
Just to give one anecdotal datapoint from Germany, we had one professor for operating systems who translated _everything_. Students disliked it a lot, basically we had to learn everything twice: the English word for day to day use/stack overflow, and the German word for this one class.
So I think translating technical terms is not a good idea in general, it might be better to explain them and use the original one then.

**Reply by mgeisler** [reply]
Yeah, I've seen this happen in Denmark too :smile: From what I understand, there was an effort to translate everything back in the 60-70'ties. So many people will talk about _EDB_ in Danish, short for _Elektronisk databehandling_, just like you have _EDV_ for _Elektronische Datenverarbeitung_ in German...

**Comment by mgeisler** [comment]
We've discussed this a bit in the past in PRs comments. My advice has been to agree on a style for your particular language. Document it somewhere: in the related issue, at the top of the PO file, etc...

I'm fine with the English words being in italic and/or in parentheses. Different languages and cultures likely have different norms here and I know that I don't have the knowledge to make editorial choices on behalf on all languages :-) The most important part  is to keep it consistent and to follow conventions, if any exist.

We could perhaps add something about this to the [style guide](https://github.com/google/comprehensive-rust/blob/main/STYLE.md)?

**Comment by deavid** [comment]
My feeling is that carefully selecting what to translate and what not is what gives the sense of quality in a translation; and the inertia of translating everything mostly comes from either automated translations or paid translators that lack domain-specific knowledge on what are they translating.

I would like to gauge here for most languages what is the most common norm for regular people (not from a language academy) and see if it can be put in the style guide as you suggest. It's probable that the consensus could be something like "don't translate technical terms in any language except for these languages: X,Y,Z".

Also I am attempting to avoid duplication of efforts. If everything gets translated forward to the target language and then it is found to be jarring and we need to go back it will need a lot of additional efforts. Instead, we can try to settle here on what could work best overall so later changes could be less impactful.

**Reply by mgeisler** [reply]
Yeah, it's a ton of work to change the style of a translation after the fact... we have ~200 pages now! So I completely agree with you that it's better to discuss this upfront.

We had a similar discussion about the use of formal vs informal form (`du` vs `Sie` in German). There the consensus was to use the information form: #818. I see this never made into the style guide, but I think it should be there.

**Reply by mgeisler** [reply]
@johnathan79717, @hueich, @wnghl, @nodmp, @superwhd, @kantasv, @jiyongp, @jooyunghan and other active translators: what did you end up doing for your respective languages?

**Reply by jiyongp** [reply]
I (mostly) followed the existing translations found in the Korean version of https://doc.rust-lang.org/book/

**Reply by kantasv** [reply]
Similar to @jiyongp, we (or mostly I) refer to the Japanese version of rustbook. I also try to keep the term as it is, or [transliterate](https://en.wikipedia.org/wiki/Transliteration) it **if**: a term is very rust-specific, or might appear in rust code. This is helpful for Japanese readers to refer back to English document when they need to. (In theory, we can completely translate those words, but it's also difficult to reconstruct original rust terms from Japanese version-> English version sometimes.) This requires judgement, of course.

**Reply by hueich** [reply]
For countries that use Traditional Chinese, most technical people are pretty comfortable with reading English, so I would opt for keeping technical terms in English to prevent having to learn those terms twice.  This also avoids the situation of different sources translating the same term differently, resulting in confusion.

**Reply by mgeisler** [reply]
@hueich, did you see my idea about adding a glossary below? I think it could be useful for English speakers as well as non-English speakers. It would show up in searches, so people could quickly find what they need and then go back to the page that covers the material in more detail.

**Reply by hueich** [reply]
Yea, I think having a glossary is helpful.  But that can work both ways, i.e. English in the material that readers can find the Chinese translation, or Chinese in the material and find the English original, so in my mind it doesn't sway the decision one way or another.  So the minor concern I have is that if an instance of a term was translated in a variation that's different than the official term in the material, readers can't find that in the glossary anymore, unless we include all the variants that we know of, which would take more work.  But I also prefer reading technical material in English, since so many existing programming resources are in English, such that I think about programming in English.  :)  So it would be helpful to get other people's input too.

**Reply by mgeisler** [reply]
> But I also prefer reading technical material in English, since so many existing programming resources are in English, such that I think about programming in English. :)

Me too :smile: I run all my computers with an English locale since it gives me the most consistent and reliable experience. So it's hard to judge how others feel about it. I live in Switzerland and I notice that people around me (outside of Google) use their machine in German — and the same in Denmark where I'm originally from. That's of course consumer products so it might not apply.

However, I have heard that there are sub-communities of developers who do all their work in their native language. Heck, I've even seen code with Danish variable names and comments once :slightly_smiling_face: It looks odd when you're used to an all English environment.

**Comment by deavid** [comment]
@vzz1x2 @JaviSorribes - as contributors/reviewers for Spanish translations, what is your take here? do you agree/disagree?

**Reply by vzz1x2** [reply]
I think that your observations are valid. I have recently seen really bad translations like "Tirantes Rizados" for Curly Brackets. 
At least in Spanish there is also the factor of Iberic Spanish vs Latin American Spanish, and even in Latin American Spanish technical terms that are used between Mexico and Argentina are different when they are translated, some times maintaining certain words in English avoids confusion there. 
We should agree on a specific style for Spanish, that in my view should try to stick to the use of English words for key technical terms.

**Reply by JaviSorribes** [reply]
I don't know about other countries, but in Spain, people do very commonly translate technical terms into Spanish in professional settings. Even though it might feel "cringe", it is quite common. So I would recommend we follow deavid's approach:

> The usual way we deal with these English terms in Spanish is to have them in italics (or quotes if italics doesn't work) then add the meaning in Spanish in parenthesis afterwards.

vzz1x2 has a good point that some translations might not be great and/or will differ from one region to another. In Spain, curly braces are called "llaves", but this might not be the case in other places. In cases where we cannot find consensus, we can add more than one term in parenthesis:

>... _curly braces_ (llaves, tirantes rizados)...

And then, to avoid pushing a perhaps not so great translation, we can follow deavid's other recommendation:

>Once the reader gets used to the new English term we just put them in italics and drop the Spanish counterpart.

The English term is bound to be more useful to a broader audience overall anyways.

**Reply by mgeisler** [reply]
> > The usual way we deal with these English terms in Spanish is to have them in italics (or quotes if italics doesn't work) then add the meaning in Spanish in parenthesis afterwards.
> 
> vzz1x2 has a good point that some translations might not be great and/or will differ from one region to another. In Spain, curly braces are called "llaves", but this might not be the case in other place.

Depending on how big the differences are, we could create translations for those dialects. I would suggest avoiding this, though, since it's a ton of work :slightly_smiling_face: If we can get away with

```markdown
foo (also known as bar in X and baz in Y)
```

in the translation, then that seems much preferable.



> The English term is bound to be more useful to a broader audience overall anyways.

Would it make sense to add a glossary to the course? Something like this:

| Term | Meaning |
| --- | --- |
| `{ ... }` | _Curly braces_ mark blocks. |
| slice | A _slice_ is a type which refers to part of an array or part of another slice. |
| `&str` | A `_string slice_`, this type gives you a read-only string. |

When translating the glossary, one could then naturally add the English words to the translation:

| Term | Mening |
| --- | --- |
| `{ ... }` | _Kryllede parenteser_ (_curly braces_ på engelsk, også kendt som _tuborgklammer_ på dansk) markerer blokke. |
| skive | En _skive_ (_slice_ på engelsk) er en type som refererer til en del af et array eller en del af en anden skive. |
| `&str` | En _streng skive_ (_string slice_ på engelsk) er en type som giver dig read-only adgang til en streng. |

**Reply by JaviSorribes** [reply]
imo a glossary would be good, but not a P0 for the course. Kind of a cherry on the top. If we did write one however, I would still suggest that we write in the following format: `<english_term> (<translated_term>)` and use the English term throughout, at least in Spanish.

**Reply by mgeisler** [reply]
A glossary could perhaps be seeded from https://doc.rust-lang.org/reference/glossary.html.

**Reply by JaviSorribes** [reply]
Reviving this discussion given https://github.com/google/comprehensive-rust/pull/1333 . In this pr the proposal is to just use the English terms in the nav and no Spanish terms. For example, in the following example we would show "Send" in the left nav: ![image](https://github.com/google/comprehensive-rust/assets/15252789/75a24724-93e2-4b36-9b62-c05dc0a1a9c6)

Is this what we want? I think this could be ok. The more I come across this, the more a glossary seems like the right way to go.

**Reply by Phosphorus-M** [reply]
I agree some keywords like Arc and Send don't make sense to translate to another language.
We can give context instead of translating the word. 
Making a glossary could be a great solution to this.

**Reply by mgeisler** [reply]
> The more I come across this, the more a glossary seems like the right way to go.

We now have a glossary: https://google.github.io/comprehensive-rust/glossary.html 🎉

You'll see it become part of the Spanish translation when you refresh the translation next with new strings.

**Reply by deavid** [reply]
It should be in English "Send", it makes no sense in Spanish. As you mention, this is a Rust specific thing and translating it adds more confusion than it helps.

**Reply by deavid** [reply]
I'm now reviewing #1333, I think that for the menus we can go with the English terms alone, as translating on the menus can add too much noise. I am leaning towards approving the PR, I don't see a problem with it.

One could argue that some folks might not know what "Send" means, but the thing is, once you click, it is explained what it means for Rust. That should be enough. If there's any case where this does not work, i.e. the original text expects the reader to understand the english meaning of the keyword, please highlight it and we can discuss each item independently.

**Reply by vzz1x2** [reply]
Yes. I agree that for the menus it works better to have the terms in English.

**Comment by mgeisler** [comment]
Just an update to people here: We now have a [glossary](https://google.github.io/comprehensive-rust/glossary.html). This is meant to be a place where you can establish the relationship between the English words and your locale.

Please help by

- filling out the English description
- adding missing words

Let me know what you think!
